outlines/generate/api.py

```diff
@@ -191,6 +197,22 @@ def __call__(
             (batch_size * num_samples), dtype=torch.float, device=self.device
         )
 
+        if include_prompt:
+            final_fsm_states = []
+            for fsm, state, prompt_token_id_seq, attention_mask_seq in mit.zip_equal(
+                fsms, fsm_states, prompt_token_ids, attention_masks):
+                assert state == 0, f"Expected state to be 0, got {state}"
+                assert isinstance(state, int), f"Expected state to be an int, got {state}"
+                for token_id, attention_mask in zip(prompt_token_id_seq, attention_mask_seq):
+                    state = fsm.get_next_state(state=state, token_id=token_id.item())
+                    if state < 0:
+                        raise ValueError(f"Invalid state {state}")
+                final_fsm_states.append(state)
```

Comment:
Would it be better to implement this as a Guide? I've started work on https://github.com/dottxt-ai/outlines/pull/531and I'd like to see Guides become more composable.  

Also `interegular` has functionality built in to make this easy.

```
>>> import interegular
>>> fsm = interegular.parse_pattern("abcd").to_fsm()
>>> fsm.accepts("abcd")
True
>>> partial_fsm = fsm.derive("ab")
>>> partial_fsm.accepts("abcd")
False
>>> partial_fsm.accepts("cd")
True
```